### PR TITLE
job isolation with sub-processes

### PR DESCRIPTION
**Pull request by theomonnom**

### CODE REVIEW

**Comment by theomonnom** [comment]
Still having some deadlock but should be ready for review while I'm working on it

File: livekit-agents/livekit/agents/ipc/client.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+    async def send(self, msg: worker.IPCJobMessage) -> None:
+        if self._closed:
+            raise Exception("ipc client is closed")
+
+        await self._send_queue.put(msg)
+
+    async def _run_conn(self) -> None:
+        process_logger.debug("connecting to ipc server")
+        rx, tx = await asyncio.open_connection("127.0.0.1", IPC_PORT)
```
Comment (keepingitneil) [inline livekit-agents/livekit/agents/ipc/client.py:109]: I haven't done a lot of research so if it's a bad idea, just ignore me. But can you use shared memory: https://docs.python.org/3.10/library/multiprocessing.shared_memory.html ?

**Review by keepingitneil — APPROVED** [review]

File: livekit-agents/livekit/agents/ipc/job_context.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+from livekit import api, rtc
+from livekit.protocol import worker, agent
+
+import asyncio
+import contextlib
+import logging
+from .client import IPCClient
+
+
+class JobContext:
```
Comment (davidzhao) [inline livekit-agents/livekit/agents/ipc/job_context.py:27]: nit: it's a bit strange why the context is located in the ipc package. two reasons:
* the context seems like a piece of code the end user would have to access frequently (I know I've had to reference it a few times when creating an agent)
* while it does communicate with parent process, it feels like that's not the core of this class

**Review by davidzhao — APPROVED** [review]
lg!

**Pull request merged**